Speed up & clean up host layout conversion

philei-tt opened the pull request:
### Ticket
Github issue: https://github.com/tenstorrent/tt-metal/issues/8621 
Previous incomplete PR: https://github.com/tenstorrent/tt-metal/pull/17848 

### Problem description
Host layout conversion is not optimal.

### What's changed
- Rewrite host layout conversion functions with better memory coalescing and preallocation of buffers.
- Add direct `LIN_ROW_MAJOR` <--> `TILED_NFACES` conversion instead of two conversions
- Remove most duplications of tilization/untilization in the code. 
- Add proper C++ test for layout conversions (different shapes, different types, different layouts)
- Add benchmark to properly compare previous and current implementations in terms of speed

#### Benchmarking:
Shape: (10240, 1024)
Biggest difference in Release is LIN_ROW_MAJOR --> TILED_NFACES (78.7ms -> 26.2ms = 3.00x speed up)
FYI: LIN_ROW_MAJOR --> TILED_NFACES == `ttnn.to_layout(row_major_tensor, ttnn.TILE_LAYOUT)`

**Release**
Before PR:
```
./build/tt_metal/benchmark/host_tilizer_untilizer/tilizer_untilizer --benchmark_min_time=10
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_ConvertLayout_RowMajorToTiledSwizzled          26.3 ms         26.3 ms          490
BM_ConvertLayout_RowMajorToTiledNfaces            78.7 ms         78.7 ms          176
BM_ConvertLayout_TiledSwizzledToRowMajor          25.5 ms         25.4 ms          529
BM_ConvertLayout_TiledSwizzledToTiledNFaces       51.9 ms         51.9 ms          268
BM_ConvertLayout_TiledNFacesToRowMajor            65.7 ms         65.7 ms          217
BM_ConvertLayout_TiledNFacesToTiledSwizzled       39.0 ms         39.0 ms          357
```

After PR: 
```
./build/tt_metal/benchmark/host_tilizer_untilizer/tilizer_untilizer --benchmark_min_time=10
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_ConvertLayout_RowMajorToTiledSwizzled          25.2 ms         25.2 ms          533
BM_ConvertLayout_RowMajorToTiledNfaces            26.2 ms         26.2 ms          552
BM_ConvertLayout_TiledSwizzledToRowMajor          25.4 ms         25.4 ms          549
BM_ConvertLayout_TiledSwizzledToTiledNFaces       27.3 ms         27.3 ms          535
BM_ConvertLayout_TiledNFacesToRowMajor            25.3 ms         25.3 ms          595
BM_ConvertLayout_TiledNFacesToTiledSwizzled       22.2 ms         22.2 ms          65
```

**Debug** -- less useful than release, but difference is huge, might be noticeable during development

Before PR:
```
./build/tt_metal/benchmark/host_tilizer_untilizer/tilizer_untilizer --benchmark_min_time=10
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_ConvertLayout_RowMajorToTiledSwizzled           145 ms          145 ms           97
BM_ConvertLayout_RowMajorToTiledNfaces             389 ms          389 ms           37
BM_ConvertLayout_TiledSwizzledToRowMajor           130 ms          130 ms          108
BM_ConvertLayout_TiledSwizzledToTiledNFaces        240 ms          240 ms           58
BM_ConvertLayout_TiledNFacesToRowMajor             278 ms          278 ms           51
BM_ConvertLayout_TiledNFacesToTiledSwizzled        145 ms          145 ms           97
```

After PR:
```
./build/tt_metal/benchmark/host_tilizer_untilizer/tilizer_untilizer --benchmark_min_time=10
--------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations
--------------------------------------------------------------------------------------
BM_ConvertLayout_RowMajorToTiledSwizzled          42.1 ms         42.1 ms          331
BM_ConvertLayout_RowMajorToTiledNfaces            55.2 ms         55.2 ms          252
BM_ConvertLayout_TiledSwizzledToRowMajor          41.3 ms         41.3 ms          397
BM_ConvertLayout_TiledSwizzledToTiledNFaces       64.5 ms         64.5 ms          170
BM_ConvertLayout_TiledNFacesToRowMajor            42.8 ms         42.8 ms          321
BM_ConvertLayout_TiledNFacesToTiledSwizzled       36.4 ms         36.4 ms          385
```

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

philei-tt committed:
Copy host tilization improvemence from #17848

philei-tt committed:
Remove (un)tilization dublication in codebase

philei-tt committed:
Rename tilization functions for clarity

philei-tt committed:
Reuse convert_layout function in (un)tilize_nfaces

philei-tt committed:
Add optimized direct host row-major -> tiled_nfaces

philei-tt committed:
Add proper layout conversion tests

philei-tt committed:
Fix compilation issues after rebase with tests branch

philei-tt committed:
Add benchmark for layout conversion

philei-tt commented:
PR is very bulky, since I update all calls to layout conversion. 
The main changes are 
- update layout conversion code: [tt_metal/impl/data_format/tilize_utils.cpp](https://github.com/tenstorrent/tt-metal/pull/20335/files#diff-f9197ee49dd22c652447ee454a6e333558df993a9ce32f0176a3014ffc8692a8), [tt_metal/api/tt-metalium/tilize_utils.hpp](https://github.com/tenstorrent/tt-metal/pull/20335/files#diff-e61fde88c7c3fa5d5e58296f74d7d1d1aa9737965acd9b29954e9eebd40254b1)
- new tests: [tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp](https://github.com/tenstorrent/tt-metal/pull/20335/files#diff-ddfaaae7224f9f1a5047428823bf3fe7ca4b0123ec95edb3293fa21f877fb2da) 
- New benchmark: [tests/tt_metal/tt_metal/benchmark/host_tilizer_untilizer/tilizer_untilizer.cpp](https://github.com/tenstorrent/tt-metal/pull/20335/files#diff-348c46c06d699c38ded2dbf65fe7a7d655ec8925f7947db21ba8ecd2deafd0d8)

philei-tt committed:
Add missing copyright

omilyutin-tt commented on tt_metal/api/tt-metalium/tilize_utils.hpp:
```diff
@@ -13,14 +13,16 @@
 
 namespace tests::utils {
 enum class TensorLayoutType {
     LIN_ROW_MAJOR = 0,   // standard element-wise row-major
     TILED_SWIZZLED = 1,  // row-major of tiles, each tile is row-major-swizzled
-    TILED_NFACES =
-        2,  // row-major of tiles, each tile is N (N = 1, 2, or 4) faces, each face is row-major, faces are swizzled
+    TILED_NFACES = 2,    // row-major of tiles, each tile is N (N = 1, 2, or 4) faces, each face is
+                         // row-major, faces are swizzled
 };
```
Unrelated to this PR, but not clear why this is in `tests::utils` namespace...

ayerofieiev-tt replied:
It all was never supposed to be used in prod code, but leaked into it..

omilyutin-tt commented on tt_metal/api/tt-metalium/tilize_utils.hpp:
```diff
@@ -13,14 +13,16 @@
+std::ostream& operator<<(std::ostream& os, TensorLayoutType layout);
 }  // namespace tests::utils
 
 using PhysicalSize = std::array<uint32_t, 2>;
```
We have Shape2D, can also be cleaned up in a follow up

omilyutin-tt commented on tests/tt_metal/tt_metal/test_matmul_multi_core_multi_dram_in0_mcast.cpp:
```diff
@@ -19,35 +19,6 @@
 using std::vector;
 using namespace tt;
 
-namespace test {
```
😄

philei-tt replied:
The amount of code duplications on test::tilize / test::untilize is remarkable ahah

omilyutin-tt commented on tests/tt_metal/tt_metal/benchmark/host_tilizer_untilizer/tilizer_untilizer.cpp:
```diff
@@ -0,0 +1,148 @@
+    std::uniform_real_distribution<float> dist(-100.0f, 100.0f);
+    for (auto& val : input_data) {
+        val = dist(gen);
+    }
```
Instead of a separate `input_data` global, can we have a method like `GetInputData` that initializes a function-local static variable (ideally using `Indestructible`), and returns the result? The next invocations would re-use the same data.

Perhaps you could then re-use `BENCHMARK_MAIN`.

omilyutin-tt replied:
Great benchmarks, btw!

philei-tt replied:
Changed to indestructible function local static variable, ptal

omilyutin-tt commented on tt_metal/impl/data_format/tilize_utils.cpp:
```diff
@@ -8,11 +8,24 @@
+        case TensorLayoutType::LIN_ROW_MAJOR: os << "LIN_ROW_MAJOR"; break;
+        case TensorLayoutType::TILED_SWIZZLED: os << "TILED_SWIZZLED"; break;
+        case TensorLayoutType::TILED_NFACES: os << "TILED_NFACES"; break;
+        default: os << "Unknown TensorLayoutType";
```
in switch cases best not to have a default - this way compiler gets broken when new enum values are added

philei-tt replied:
Will remove

omilyutin-tt commented on tt_metal/impl/data_format/tilize_utils.cpp:
```diff
@@ -275,6 +273,141 @@ std::vector<T> convert_to_flat_layout(
+    uint32_t W = shape[1];
+    uint32_t B = in_row_major.size() / (H * W);
+
+    TT_ASSERT(in_row_major.size() > 0 and H > 0 and W > 0, "None of the input size, H, nor W can be 0");
+    TT_ASSERT((in_row_major.size() % (H * W)) == 0, "Input size must be divisible by H and W");
```
Can be fatal? Since this is user input?

philei-tt replied:
Yep, makes sense. I will change

omilyutin-tt commented on tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp:
```diff
@@ -8,96 +8,493 @@
 }
+}  // namespace reference
+
+using Type = bfloat16;
```
Is this useful?

philei-tt replied:
What do you mean? But yeah, I wanted to test multiple types, but seems like I forgot to add this to tests

omilyutin-tt commented on tests/tt_metal/tt_metal/api/test_tilize_untilize.cpp:
```diff
@@ -8,96 +8,493 @@
+
+using Type = bfloat16;
+
+// Note: tuple is used for ::testing::Combine
```
Can we use https://github.com/google/googletest/blob/main/googletest/include/gtest/gtest-param-test.h#L411-L447 here?

philei-tt replied:
Yes, I will try

philei-tt replied:
Yeah, it's kind of tricky to parametrize over regular parameters and types at the same time. I did the trick with lambda inside each test that takes a type as argument, and manually run for all types. Hope it's good enough

omilyutin-tt replied:
Yeah, it is annoying to use these advanced features together, often simply doesn't work.

`ConvertGenerator` though allows use to get rid of std::tuple in the interface, so `TilizeUntilizeParams` can be a struct with named fields; you then use `Combine` to generate the tuples that are converted to that named struct. Maybe too convoluted to do here, but maybe worth a try

omilyutin-tt replied:
W.r.t. parameterization, https://github.com/tenstorrent/tt-metal/blob/9410ecfe84778997e5278ada8603295961b895c8/tests/ttnn/unit_tests/gtests/tensor/test_vector_conversion.cpp is another data point, in case this is useful. I chose to iterate over params and used `TYPED_TEST` there, went the other way.

philei-tt committed:
Remove tests::utils namespace around TensorLayoutType.

Also, use function-local static indestructible for data gen in tilize
benchmark

philei-tt committed:
Impprove input validation and change assert->fatal in convert_xxx

philei-tt committed:
Use func-local static data instead of static class member in test

philei-tt committed:
Test bfloat16, float, int32, uint32 in test_tilize_untilize.cpp

philei-tt committed:
Fix a few broken tests

philei-tt committed:
Fix case when tile_shape == face_shape

philei-tt committed:
Add proper support of transpose_face_order

philei-tt committed:
Fix support for transpose_within_face and face_shape==tile_shape

philei-tt committed:
Fix compilation issue after rebase

philei-tt committed:
Fix case of non-square tiles

philei-tt committed:
Fix compilation after rebase

philei-tt committed:
Fix all broken cases for N_FACES->ROW_MAJOR

philei-tt committed:
Fix includes in test_tilize_untilize.cpp
